ReactiveCocoa/UIKit/iOS/UISearchBar.swift

```diff
@@ -2,27 +2,68 @@ import ReactiveSwift
+	@objc func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
+		forwardee?.searchBarTextDidEndEditing?(searchBar)
+	}
 
-	@objc func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {}
-	@objc func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {}
+	@objc func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+		forwardee?.searchBar?(searchBar, textDidChange: searchText)
+	}
 }
 
-private var delegateKey: UInt8 = 0
+private let hasSwizzledKey = AssociationKey<Bool>(default: false)
+private let proxyKey = AssociationKey<SearchBarDelegateProxy?>()
 
 extension Reactive where Base: UISearchBar {
 
-	private var delegate: ReactiveUISearchBarDelegate {
-		if let delegate = objc_getAssociatedObject(base, &delegateKey) as? ReactiveUISearchBarDelegate {
-			return delegate
-		} else if let _ = base.delegate {
-			fatalError("Cannot use reactive values on UISearchBar with a custom delegate!")
-		}
+	private var proxy: SearchBarDelegateProxy {
+		return base.synchronized {
+			if let proxy = base.associations.value(forKey: proxyKey) {
+				return proxy
+			}
+
+			let subclass: AnyClass = swizzleClass(base)
+
+			// Hide the original setter, and redirect subsequent delegate assignment
+			// to the proxy.
+			synchronized(subclass) {
+				let subclassAssociations = Associations(subclass as AnyObject)
+
+				if !subclassAssociations.value(forKey: hasSwizzledKey) {
+					subclassAssociations.setValue(true, forKey: hasSwizzledKey)
+
+					let method = class_getInstanceMethod(subclass, #selector(setter: base.delegate))
 
-		let delegate = ReactiveUISearchBarDelegate()
-		base.delegate = delegate
-		objc_setAssociatedObject(base, &delegateKey, delegate, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
-		return delegate
+					let newImpl: @convention(block) (NSObject, UISearchBarDelegate) -> Void = { object, delegate in
+						let proxy = object.associations.value(forKey: proxyKey)!
+						proxy.forwardee = delegate
+					}
+
+					class_replaceMethod(subclass,
+					                    #selector(setter: base.delegate),
+					                    imp_implementationWithBlock(newImpl as Any),
+					                    method_getTypeEncoding(method))
+				}
+			}
+
+			// Set the proxy as the delegate.
+			let realClass: AnyClass = class_getSuperclass(subclass)
+			let impl = class_getMethodImplementation(realClass, #selector(setter: base.delegate))
+			typealias OriginalImpl = @convention(c) (AnyObject, Selector, AnyObject) -> Void
+
+			let proxy = SearchBarDelegateProxy { [weak base] proxy in
+				guard let base = base else { return }
+				unsafeBitCast(impl, to: OriginalImpl.self)(base, #selector(setter: base.delegate), proxy)
+			}
+
+			base.associations.setValue(proxy, forKey: proxyKey)
+
+			proxy.forwardee = base.delegate
+			unsafeBitCast(impl, to: OriginalImpl.self)(base, #selector(setter: base.delegate), proxy)
+
+			return proxy
```

Comment:
It seems like we'd want to reuse this for other controls with delegates. Maybe we should extract this somewhere? Maybe event to `DelegateProxy`?